README.md
# php7parser

A fast and forgiving PHP7 recursive descent parser implemented in Typescript. 

The parser outputs a parse tree of phrases (branches) and tokens (leaves). 
The full source code is represented by the tree. Trivia is accessible via `Token.previous`.

## Design Goals

* Modern browser and nodejs compatibility.
* Error tolerant and high performance.
* Output representative of full source code.
* Adherance to the PHP language specifications.
* Prefer error tolerance over enforcement of language constraints.

## Usage

```typescript
    import { Parser } from 'php7parser';

    let src = '<?php echo "Hello World!";';
    let tree = Parser.parse(src);
```

## Interface

```typescript

    export declare namespace Parser {
        function parse(text: string): Phrase;
    }

    export interface Node {
        kind: number;
        length: number;
    }

    export interface Phrase extends Node {
        /**
        * Phrase and token child nodes
        */
        children: Node[];
    }

    export interface ParseError extends Phrase {

        /**
        * The token that prompted the parse error
        */
        unexpected: Token;

        /**
        * The expected token type
        */
        expected?: TokenKind;

    }

    export interface Token extends Node {
        /**
        * Offset within source were first char of token is found
        * @deprecated
        */
        offset: number,
        /**
        * The previous token
        */
        previous: Token
    }

```


CHANGELOG.md
# Change Log

## [2.0.0] - TBA
#### Added
- 7.3 features
- TopStatementList andScript kinds
- length property to tokens and phrases
- previous property to tokens

#### Changed
- Skip {} subscript contents on parse error
- Removed additional error phrase kinds
- Removed trivia tokens from tree. Now accessible via Token.previous.
- Renamed tokenType and phraseType to kind
- Removed modestack from token
- Revert 1.0.8 dont skip src after halt compiler
- Deprecated Token.offset

## [1.0.8] - 2018-02-10
#### Fixed
- Treat OpenTagEcho as Echo
- Dont skip src after halt compiler stmt
- Add expected to parse error

## [1.0.7] - 2018-01-05
#### Fixed
- Inf. loop on merge conflict marker

## [1.0.6] - 2017-12-28
#### Fixed
- `!==` length

## [1.0.5] - 2017-12-16
#### Changed
- Allow utf8 names to match official php lexer implementation.

## [1.0.4] - 2017-12-09
#### Changed
- Dynamic var use encaps expr child

## [1.0.3] - 2017-12-02
#### Fixed
- Incorrectly reporting parse error on anon. function expression statement.

## [1.0.2] - 2017-11-01
#### Fixed
- Close tag on empty inline comment not being recognised 

## [1.0.1] - 2017-10-11
#### Fixed
- Endless loop on error inside namespace use decl when namespace keyword is alias
- Trailing whitespace missing from parse tree

## [1.0.0] - 2017-07-27
#### Changed
- Refactored lexer. No longer uses reg. exp.
- Up to 4x improved performance
- Removed interface declarations for each node type
- Parse errors now reported in tree as own nodes with skipped child tokens.
#### Fixed
- Single line comments should not consume trailing whitespace

## [0.9.9]- 2017-07-03
#### Fixed
- Missing unset cast in precedence table

## [0.9.8] - 2017-06-22
#### Fixed
- Whitespace inside on namespace use decl

## [0.9.7] - 2017-06-20
#### Fixed
- ArrayInitialiser allow empty element
- Floats with exponent
- Allow optional expr with break and continue

## [0.9.6] - 2017-06-20
#### Fixed
- Complex string syntax
- ConstantAccessExpression
- Errors not reported on ArrayInitialiserList

## [0.9.4] - 2017-04-23
#### Fixed
- Use group declaration missing }
- Close heredoc lexer bug 

## [0.9.3] - 2017-04-21
#### Fixed
- Use trait mangles parse tree

## [0.9.2] - 2017-04-21
#### Fixed
- Check for expect() null return.

## [0.9.1] - 2017-04-17
#### Fixed
- Wrong type used for class constant element.

## [0.9.0] - 2017-04-16
Initial release


lib/lexer.d.ts
export declare const enum TokenType {
    Unknown = 0,
    EndOfFile = 1,
    Abstract = 2,
    Array = 3,
    As = 4,
    Break = 5,
    Callable = 6,
    Case = 7,
    Catch = 8,
    Class = 9,
    ClassConstant = 10,
    Clone = 11,
    Const = 12,
    Continue = 13,
    Declare = 14,
    Default = 15,
    Do = 16,
    Echo = 17,
    Else = 18,
    ElseIf = 19,
    Empty = 20,
    EndDeclare = 21,
    EndFor = 22,
    EndForeach = 23,
    EndIf = 24,
    EndSwitch = 25,
    EndWhile = 26,
    EndHeredoc = 27,
    Eval = 28,
    Exit = 29,
    Extends = 30,
    Final = 31,
    Finally = 32,
    For = 33,
    ForEach = 34,
    Function = 35,
    Global = 36,
    Goto = 37,
    HaltCompiler = 38,
    If = 39,
    Implements = 40,
    Include = 41,
    IncludeOnce = 42,
    InstanceOf = 43,
    InsteadOf = 44,
    Interface = 45,
    Isset = 46,
    List = 47,
    And = 48,
    Or = 49,
    Xor = 50,
    Namespace = 51,
    New = 52,
    Print = 53,
    Private = 54,
    Public = 55,
    Protected = 56,
    Require = 57,
    RequireOnce = 58,
    Return = 59,
    Static = 60,
    Switch = 61,
    Throw = 62,
    Trait = 63,
    Try = 64,
    Unset = 65,
    Use = 66,
    Var = 67,
    While = 68,
    Yield = 69,
    YieldFrom = 70,
    DirectoryConstant = 71,
    FileConstant = 72,
    LineConstant = 73,
    FunctionConstant = 74,
    MethodConstant = 75,
    NamespaceConstant = 76,
    TraitConstant = 77,
    StringLiteral = 78,
    FloatingLiteral = 79,
    EncapsulatedAndWhitespace = 80,
    Text = 81,
    IntegerLiteral = 82,
    Name = 83,
    VariableName = 84,
    Equals = 85,
    Tilde = 86,
    Colon = 87,
    Semicolon = 88,
    Exclamation = 89,
    Dollar = 90,
    ForwardSlash = 91,
    Percent = 92,
    Comma = 93,
    AtSymbol = 94,
    Backtick = 95,
    Question = 96,
    DoubleQuote = 97,
    SingleQuote = 98,
    LessThan = 99,
    GreaterThan = 100,
    Asterisk = 101,
    AmpersandAmpersand = 102,
    Ampersand = 103,
    AmpersandEquals = 104,
    CaretEquals = 105,
    LessThanLessThan = 106,
    LessThanLessThanEquals = 107,
    GreaterThanGreaterThan = 108,
    GreaterThanGreaterThanEquals = 109,
    BarEquals = 110,
    Plus = 111,
    PlusEquals = 112,
    AsteriskAsterisk = 113,
    AsteriskAsteriskEquals = 114,
    Arrow = 115,
    OpenBrace = 116,
    OpenBracket = 117,
    OpenParenthesis = 118,
    CloseBrace = 119,
    CloseBracket = 120,
    CloseParenthesis = 121,
    QuestionQuestion = 122,
    Bar = 123,
    BarBar = 124,
    Caret = 125,
    Dot = 126,
    DotEquals = 127,
    CurlyOpen = 128,
    MinusMinus = 129,
    ForwardslashEquals = 130,
    DollarCurlyOpen = 131,
    FatArrow = 132,
    ColonColon = 133,
    Ellipsis = 134,
    PlusPlus = 135,
    EqualsEquals = 136,
    GreaterThanEquals = 137,
    EqualsEqualsEquals = 138,
    ExclamationEquals = 139,
    ExclamationEqualsEquals = 140,
    LessThanEquals = 141,
    Spaceship = 142,
    Minus = 143,
    MinusEquals = 144,
    PercentEquals = 145,
    AsteriskEquals = 146,
    Backslash = 147,
    BooleanCast = 148,
    UnsetCast = 149,
    StringCast = 150,
    ObjectCast = 151,
    IntegerCast = 152,
    FloatCast = 153,
    StartHeredoc = 154,
    ArrayCast = 155,
    OpenTag = 156,
    OpenTagEcho = 157,
    CloseTag = 158,
    Comment = 159,
    DocumentComment = 160,
    Whitespace = 161,
}
export declare const enum LexerMode {
    Initial = 0,
    Scripting = 1,
    LookingForProperty = 2,
    DoubleQuotes = 3,
    NowDoc = 4,
    HereDoc = 5,
    EndHereDoc = 6,
    Backtick = 7,
    VarOffset = 8,
    LookingForVarName = 9,
}
export interface Token {
    /**
     * Token type
     */
    tokenType: TokenType;
    /**
     * Offset within source were first char of token is found
     */
    offset: number;
    /**
     * Length of token string
     */
    length: number;
    /**
     * Lexer mode prior to this token being read.
     */
    modeStack: LexerMode[];
}
export declare namespace Lexer {
    function setInput(text: string, lexerModeStack?: LexerMode[], lastPosition?: number): void;
    function lex(): Token;
}
export declare function tokenTypeToString(type: TokenType): "Unknown" | "EndOfFile" | "Abstract" | "Array" | "As" | "Break" | "Callable" | "Case" | "Catch" | "Class" | "ClassConstant" | "Clone" | "Const" | "Continue" | "Declare" | "Default" | "Do" | "Echo" | "Else" | "ElseIf" | "Empty" | "EndDeclare" | "EndFor" | "EndForeach" | "EndIf" | "EndSwitch" | "EndWhile" | "EndHeredoc" | "Eval" | "Exit" | "Extends" | "Final" | "Finally" | "For" | "ForEach" | "Function" | "Global" | "Goto" | "HaltCompiler" | "If" | "Implements" | "Include" | "IncludeOnce" | "InstanceOf" | "InsteadOf" | "Interface" | "Isset" | "List" | "And" | "Or" | "Xor" | "Namespace" | "New" | "Print" | "Private" | "Public" | "Protected" | "Require" | "RequireOnce" | "Return" | "Static" | "Switch" | "Throw" | "Trait" | "Try" | "Unset" | "Use" | "Var" | "While" | "Yield" | "YieldFrom" | "DirectoryConstant" | "FileConstant" | "LineConstant" | "FunctionConstant" | "MethodConstant" | "NamespaceConstant" | "TraitConstant" | "StringLiteral" | "FloatingLiteral" | "EncapsulatedAndWhitespace" | "Text" | "IntegerLiteral" | "Name" | "VariableName" | "Equals" | "Tilde" | "Colon" | "Semicolon" | "Exclamation" | "Dollar" | "ForwardSlash" | "Percent" | "Comma" | "AtSymbol" | "Backtick" | "Question" | "DoubleQuote" | "SingleQuote" | "LessThan" | "GreaterThan" | "Asterisk" | "AmpersandAmpersand" | "Ampersand" | "AmpersandEquals" | "CaretEquals" | "LessThanLessThan" | "LessThanLessThanEquals" | "GreaterThanGreaterThan" | "GreaterThanGreaterThanEquals" | "BarEquals" | "Plus" | "PlusEquals" | "AsteriskAsterisk" | "AsteriskAsteriskEquals" | "Arrow" | "OpenBrace" | "OpenBracket" | "OpenParenthesis" | "CloseBrace" | "CloseBracket" | "CloseParenthesis" | "QuestionQuestion" | "Bar" | "BarBar" | "Caret" | "Dot" | "DotEquals" | "CurlyOpen" | "MinusMinus" | "ForwardslashEquals" | "DollarCurlyOpen" | "FatArrow" | "ColonColon" | "Ellipsis" | "PlusPlus" | "EqualsEquals" | "GreaterThanEquals" | "EqualsEqualsEquals" | "ExclamationEquals" | "ExclamationEqualsEquals" | "LessThanEquals" | "Spaceship" | "Minus" | "MinusEquals" | "PercentEquals" | "AsteriskEquals" | "Backslash" | "BooleanCast" | "UnsetCast" | "StringCast" | "ObjectCast" | "IntegerCast" | "FloatCast" | "StartHeredoc" | "ArrayCast" | "OpenTag" | "OpenTagEcho" | "CloseTag" | "Comment" | "DocumentComment" | "Whitespace";
